### ISSUE TITLE
[inductor] SDPA fusion matches a permuted K as if it were key.transpose(-2,-1), silently producing wrong attention outputs

### ISSUE THREAD

**Issue by laolvfan** [pytorch/pytorch#195320]
## 🐛 Describe the bug

Inductor's SDPA fusion pass rewrites a multi-head attention module into `scaled_dot_product_attention` when it should not, and the compiled model silently returns wrong numbers. There is no error and no warning — eager and compiled just disagree.

The trigger is a common hand-written MHA optimisation: permuting `K` straight into the transposed layout `(B, H, D, T)` so the score matmul needs no `transpose(-2, -1)`:

```python
k = self.k(x).view(b, t, H, D).permute(0, 2, 3, 1)   # (B, H, D, T)
w = (torch.matmul(q, k) * scale).softmax(dim=-1)     # no transpose needed
```

The patterns in `torch/_inductor/fx_passes/fuse_attention.py` are written as

```python
def _sfdp_pattern_2(query, key, value, scale_factor):
    return (
        torch.matmul(query, key.transpose(-2, -1))
        .mul(scale_factor)
        .softmax(dim=-1)
        .matmul(value)
    )
```

but the matcher accepts **any** dimension permutation where `key.transpose(-2, -1)` is expected, and binds the *pre-permute* tensor as `key`. On a 4-D tensor `permute(0, 2, 3, 1)` also moves the batch axes, so it is a different computation — yet it matches. The replacement then receives a `query`/`key`/`value` triple that does not correspond to the original graph.

## Reproduction

```python
import torch
import torch.nn as nn

B, T, H, D = 1, 8, 8, 8          # seq_len == n_heads
E = H * D

class MHA(nn.Module):
    """Multi-head attention where K is permuted straight into the
    transposed layout (B, H, D, T), so the score matmul needs no
    transpose(-2, -1). This is a common hand-written optimisation."""
    def __init__(self):
        super().__init__()
        self.q = nn.Linear(E, E, bias=False)
        self.k = nn.Linear(E, E, bias=False)
        self.v = nn.Linear(E, E, bias=False)

    def forward(self, x):
        b, t, _ = x.shape
        q = self.q(x).view(b, t, H, D).permute(0, 2, 1, 3)   # (B, H, T, D)
        k = self.k(x).view(b, t, H, D).permute(0, 2, 3, 1)   # (B, H, D, T) pre-transposed
        v = self.v(x).view(b, t, H, D).permute(0, 2, 1, 3)   # (B, H, T, D)
        w = (torch.matmul(q, k) * (D ** -0.5)).softmax(dim=-1)
        return (w @ v).permute(0, 2, 1, 3).reshape(b, t, E)

torch.manual_seed(0)
m = MHA().eval()
x = torch.randn(B, T, E)

with torch.no_grad():
    eager = m(x)
    compiled = torch.compile(m, backend="inductor")(x)

print("allclose:", torch.allclose(eager, compiled, atol=1e-4, rtol=1e-3))
print("max abs diff:", (eager - compiled).abs().max().item())
```

## Actual result

```
allclose: False
max abs diff: 0.4378347098827362
```

Identical on 2.11.0 and 2.13.0. There are no random ops in the model and the compiled function is deterministic (calling it twice gives identical output), so this is not RNG divergence; the magnitude is far past any precision effect.

## Evidence that the fusion pass is responsible

1. **eager is correct.** Recomputing the same expression with plain tensor ops (no module, no compile) matches eager exactly on every seed tested.

2. **Disabling the pattern matcher fixes it.** With `torch._inductor.config.pattern_matcher = False`, compiled matches eager to `1.19e-07`.

3. **Control — the standard spelling is fine.** Replacing the pre-transposed `K` with the textbook `torch.matmul(q, k.transpose(-2, -1))` (same shapes, same replacement function still fires) gives `match=True`, `maxdiff=8.9e-08`.

4. **Instrumenting the replacement shows the mis-binding.** Wrapping each `_sfdp_replacement_N` to log its arguments shows `key` bound to the pre-permute tensor rather than the operand actually feeding the matmul.

## Shape sweep

Same MHA module, varying shape. `T == H` silently miscomputes; otherwise it crashes (that crash is #195282).

| B | T | H | D | torch 2.11.0 | torch 2.13.0 |
|---:|---:|---:|---:|---|---|
| 2 | 16 | 4 | 8 | OK (not matched) | **crash** (`_replacement_28`) |
| 1 | 8 | 8 | 8 | **wrong**, 0.4378 (`_replacement_2`) | **wrong**, 0.4378 |
| 2 | 4 | 4 | 16 | OK (not matched) | **wrong**, 0.5083 (`_replacement_28`) |
| 1 | 128 | 8 | 64 | **crash** (`_replacement_2`) | **crash** |
| 4 | 32 | 4 | 32 | OK (not matched) | **crash** (`_replacement_28`) |
| 1 | 4 | 4 | 4 | **wrong**, 0.2538 (`_replacement_2`) | **wrong**, 0.2538 |

Two things worth noting:

- The bug is **not new** — `_sfdp_replacement_2` already mis-matches on 2.11.0.
- `_sfdp_pattern_28` ("Visformer pattern — same as pattern 4 but non-contiguous q/k/v") does not exist in 2.11.0. Adding it **widened the blast radius**: three shape configurations that compiled correctly on 2.11.0 now crash or silently miscompute on 2.13.0.

## Versions

- PyTorch `2.11.0` — reproduces
- PyTorch `2.13.0` — reproduces (identical value)

CPU only, Linux x86_64, Python 3.11.

## Expected behavior

The pattern should only match when the tensor bound to `key` genuinely reaches the matmul through a last-two-dims transpose, or the pass should verify that the captured `query`/`key`/`value` reproduce the matched subgraph before substituting `scaled_dot_product_attention`. A permutation that moves batch axes is a different computation and must not be fused as attention.

## Additional context

I reported the crashing manifestation of this same matcher flaw in #195282. Filing the silent-wrong-result case separately since the severity is different — a crash is visible, this one silently corrupts outputs.

Originally found by differential testing (eager vs `torch.compile`) over StarCoder-generated programs targeting Inductor's `sfdp` fusion patterns; the MHA module above is a realistic reduction of what those programs were exercising.

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by KHARSHAVARDHAN-eng** [comment]
@laolvfan , Hi! I'd like to work on this issue. I’ll start by reproducing the problem and investigating the relevant code path. If there are no objections, I’d be happy to submit a fix along with appropriate regression tests.

**Comment by laolvfan** [comment]
Hi @KHARSHAVARDHAN-eng. Thanks for picking this up — no objections at all.

  One thing that may help scope the fix: the mis-match is not specific to _sfdp_pattern_2. Applying the same permuted-K spelling across the family also triggers _sfdp_replacement_1, _5 and _28, and #195282 is the crashing manifestation of the same root cause. A fix that only guards one pattern would likely leave the others. The control experiment (proper transpose(-2,-1) compiles fine and matches eager) is in the issue body and should make a regression test straightforward.
